File: modules/roaming-guard/src/RoamingGuardModule.ts
import { NativeModule, requireNativeModule } from "expo";

export type RoamingZone = {
  id: string;
  name: string;
  countries: string[];
};

declare class RoamingGuardModule extends NativeModule<{}> {
  getAllowedCountries(): Promise<string[]>;
  getStandaloneCountries(): Promise<string[]>;
  getSelectedZoneIds(): Promise<string[]>;
  addCountry(countryCode: string): Promise<void>;
  removeCountry(countryCode: string): Promise<void>;
  isCountryAllowed(countryCode: string): Promise<boolean>;
  addZone(zoneId: string): Promise<void>;
  removeZone(zoneId: string): Promise<void>;
  getAvailableZones(): RoamingZone[];
}

export default requireNativeModule<RoamingGuardModule>("RoamingGuard");

File: modules/roaming-guard/android/src/main/java/expo/modules/roamingguard/RoamingGuardModule.kt
package expo.modules.roamingguard

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class RoamingGuardModule : Module() {

    private val countryStorage: CountryStorage
        get() {
            val context = requireNotNull(appContext.reactContext) {
                "React context not available"
            }

            return CountryStorage(context.applicationContext)
        }

    override fun definition() = ModuleDefinition {

        Name("RoamingGuard")

        val getAllowedCountriesCoroutine:
            suspend () -> List<String> = {
                countryStorage.getAllowedCountries()
            }

        AsyncFunction("getAllowedCountries") Coroutine
            getAllowedCountriesCoroutine

        AsyncFunction("addCountry") Coroutine { countryCode: String ->
            countryStorage.addCountry(countryCode)
        }

        AsyncFunction("removeCountry") Coroutine { countryCode: String ->
            countryStorage.removeCountry(countryCode)
        }

        AsyncFunction("isCountryAllowed") Coroutine { countryCode: String ->
            countryStorage.isCountryAllowed(countryCode)
        }
    }
}

